File: kth-smallest-sum-of-matrix.py
from heapq import heappush, heappop

class Solution:
    def kthSmallest(self, mat: list[list[int]], k: int) -> int:
        if not mat:
            return 0
        
        COLS = len(mat[0])
        
        heap = [(sum([row[0] for row in mat]), [0 for _ in mat])]
        visited = set() # holds maximum k tuples of length m
        ret = []
        
        while heap and len(ret) < k: # O(k lg k)
            s, idx_lst = heappop(heap)
            ret.append(s)
            visited.add(tuple(idx_lst))
			
            for row, col in enumerate(idx_lst): # O(m)
                # print(row, col)
                idx_lst[row] = col + 1
                print(idx_lst)
                idx_lst_tup = tuple(idx_lst) # O(m)
				
                if col + 1 < COLS and idx_lst_tup not in visited:
                    new_s = s - mat[row][col] + mat[row][col + 1]
                    heappush(heap, (new_s, idx_lst.copy()))
                    visited.add(idx_lst_tup)
					
                idx_lst[row] = col
            
        return ret[-1]

mat = [
[13, 48, 86, 76, 70],
[36, 95, 31, 80, 56],
[46, 1 ,93 ,92 ,61],
[66, 94, 76, 15, 41],
[93, 33, 53, 97, 89]
]
abc = Solution()
print(abc.kthSmallest(mat, 5))

File: kth-smallest-sum-of-matrix.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll long long

class Solution 
{
    public : int kthSmallest(vector<vector<int>> &mat, int k)
    {
        int m = mat.size();
        int n = mat[0].size();
        priority_queue<vector<int>> pq;
        int sum = 0;
        vector<int> temp(m + 1, 0);
        for (int i = 0; i < m; i++)
        {
            sum += mat[i][0];
        }
        temp[0] =-(sum);
        pq.push(temp);
        int count = 1;
        map<vector<int>, int> mp;
        mp[temp] = 1;
        while (!pq.empty())
        {
            if (count == k)
                break;
            temp = pq.top();
            pq.pop();
            for (int i = 0; i < m; i++)
            {
                vector<int> t = temp;
                if (t[i + 1] < n - 1)
                {
                    int val = -t[0];
                    val -= mat[i][t[i + 1]];
                    val += mat[i][t[i + 1] + 1];
                    t[i + 1]++;
                    t[0] = -val;

                    if (mp[t] != 1)
                    {
                        pq.push(t);
                        mp[t] = 1;
                    }
                }
            }
            count++;
        }
        temp = pq.top();
        return -temp[0];
    }
};